File: _Src/Freem/Freem.Web.Api.Public.Client/Implementations/TagsClient.cs
using System.Text.Json;
using Freem.Entities.Tags.Identifiers;
using Freem.Entities.Tags.Models;
using Freem.Http.Requests.Abstractions;
using Freem.Http.Requests.Entities;
using Freem.Http.Requests.Entities.Extensions;
using Freem.Web.Api.Public.Client.Implementations.Base;
using Freem.Web.Api.Public.Client.Models;
using Freem.Web.Api.Public.Contracts.Tags;

namespace Freem.Web.Api.Public.Client.Implementations;

public sealed class TagsClient : BaseClient
{
    private readonly JsonSerializerOptions _options;
    
    public TagsClient(IHttpClient client, JsonSerializerOptions options, TokenLoader tokenLoader) 
        : base(client, options, tokenLoader)
    {
        ArgumentNullException.ThrowIfNull(options);
        
        _options = options;
    }

    public Task<ClientResult<CreateTagResponse>> CreateAsync(CreateTagRequest body, CancellationToken cancellationToken = default)
    {
        var request = HttpRequest.Post("api/v1/tags")
            .WithJsonBody(body, _options);

        return SendAuthorizedAsync<CreateTagResponse>(request, cancellationToken);
    }

    public Task<ClientResult> UpdateAsync(
        TagIdentifier id,
        UpdateTagRequest body, 
        CancellationToken cancellationToken = default)
    {
        var request = HttpRequest.Put($"api/v1/tags/{id}")
            .WithJsonBody(body, _options);

        return SendAuthorizedAsync(request, cancellationToken);
    }

    public Task<ClientResult> RemoveAsync(TagIdentifier id, CancellationToken cancellationToken = default)
    {
        var request = HttpRequest.Delete($"api/v1/tags/{id}");
        
        return SendAuthorizedAsync(request, cancellationToken);
    }

    public Task<ClientResult<TagResponse>> GetAsync(TagIdentifier id, CancellationToken cancellationToken = default)
    {
        var request = HttpRequest.Get($"api/v1/tags/{id}");
        
        return SendAuthorizedAsync<TagResponse>(request, cancellationToken);
    }

    public Task<ClientResult<TagResponse>> GetAsync(TagName name, CancellationToken cancellationToken = default)
    {
        var request = HttpRequest.Get($"api/v1/tags/by-name/{name}");
        
        return SendAuthorizedAsync<TagResponse>(request, cancellationToken);
    }

    public Task<ClientResult<IAsyncEnumerable<TagResponse>>> ListAsync(
        ListTagRequest query, CancellationToken cancellationToken = default)
    {
        var request = HttpRequest.Get($"api/v1/tags")
            .WithQueryParameter(nameof(query.Limit), query.Limit.ToString())
            .WithQueryParameter(nameof(query.Offset), query.Offset.ToString());
        
        return SendAuthorizedAsync<IAsyncEnumerable<TagResponse>>(request, cancellationToken);
    }
}

File: _Src/Freem/Freem.Web/src/clients/TagsClient.ts
import {CreateTagRequest} from "./models/tags/CreateTagRequest.ts";
import {UpdateTagRequest} from "./models/tags/UpdateTagRequest.ts";
import {ListTagRequest} from "./models/tags/ListTagRequest.ts";
import {IHttpClient} from "./http/IHttpClient.ts";

export class TagsClient {
  private readonly client: IHttpClient;

  constructor(client: IHttpClient) {
    this.client = client;
  }

  public async create(request: CreateTagRequest): Promise<Response> {
    return await this.client.post({url: "api/v1/tags", body: request});
  }

  public async update(id: string, request: UpdateTagRequest): Promise<Response> {
    return await this.client.post({url: `api/v1/tags/${id}`, body: request});
  }

  public async remove(id: string): Promise<Response> {
    return await this.client.delete({url: `api/v1/tags/${id}`});
  }

  public async get(id: string): Promise<Response> {
    return await this.client.get({url: `api/v1/tags/${id}`});
  }

  public async findByName(name: string): Promise<Response> {
    return await this.client.get({url: `api/v1/tags/by-name/${name}`});
  }

  public async list(request: ListTagRequest): Promise<Response> {
    return await this.client.get({url: "api/v1/tags", query: request});
  }
}